### ISSUE TITLE
Configurable limits/circuit breakers on mesh-gateways

### ISSUE THREAD

**Issue by danieleva** [hashicorp/consul#13589]
#### Feature Description

Mesh-gateways envoy is configured with a cluster for each federated datacenter, so the default circuit breaker config applies with a max 1024 concurrent connections. Limits can be configured for upstreams on sidecar proxies but there's no equivalent for mesh-gateways. It would be awesome to be able to configure the limits for max connections towards a federated datacenter.

#### Use Case(s)
With enough services spread across federated datacenters it's easy to reach the 1024 concurrent connection limit. Scaling mesh-gateways horizontally helps but forces us to run an overprovisioned infrastructure, mostly idle as all nodes are blocked when they reach 1024 connections, even if the gateways have plenty of resources available.

**Comment by Amier3** [comment]
Hey @danieleva 

Thanks for making this, setting max connections at various levels has come up a couple of times lately  ( #13374 & #12373 ), so this is something that's actively on our radar to fix.

**Comment by Hillkorn** [comment]
Is there any workaround to configure the limits except scaling mesh gateway instances

**Comment by danieleva** [comment]
The only option I found until a config toggle is exposed was to patch our internal consul build and hardcode an arbitrary value that works for our use cases.

**Comment by huikang** [comment]
I am looking into the implementation of this feature. Based on the current configurable fields in service mesh, there seems to be 2 options to enable this feature:

* Option 1

Use service-defaults to set the default limit for all the upstream services of the mesh gateway service and override the default value in exported-service config entry, e.g.,

```
Kind = "service-defaults"
Name = "mesh-gateway"

UpstreamConfig = {
  Defaults = {
    Limits = {
      MaxConnections = 512         # <- new; default for upstream svc
      MaxPendingRequests = 512     # <- new; default for upstream svc
      MaxConcurrentRequests = 512  # <- new; default for upstream svc
    }
  }
}
```

Override the mesh gateway for service `foo`:
```
Kind = "exported-services"
Name = "default"
Services = [
  {
    Name      = "foo"
    Namespace = "default"
    Consumers = [
      {
        Peer = "peering-cluster"
      }
    ]
    MaxConnections = 2048            // <- new
    MaxPendingRequests = 4096        // <- new
    MaxConcurrentRequests = 3000     // <- new
  }
]
```

(what if the limit of same exported service are set differently for different consumers?)

* Option 2

Add a new config entry “Kind = "mesh-gateway", which is similar to the ingress-gateway config entry

```
Kind = "mesh-gateway"
Name = "mesh-gateway"

Defaults {
  MaxConnections = 4096    // default for all upstream service
}

UpstreamConfig = [
  {
    Services = [
      {
        Name = "foo"
        MaxConnections = 2048  // override the default value 
        Consumers = [
          {
            Peer = "peering-cluster"
          }
        ]
      }
    ]
  }
]
```

@jkirschner-hashicorp, @mkeeler , what do you think? Thanks.

**Comment by Hillkorn** [comment]
Any Update on this issue?

**Comment by sarahalsmiller** [comment]
Good morning, Daniele. We are taking a look at implementing this, and had a question about your use case. Do you need to configure the limits per service or would it be acceptable to configure the limits at the gateway level?

**Comment by danieleva** [comment]
It will be enough to be able to configure at the gateway level, no need to have service level granularity. Thank you :D

**Comment by david-yu** [comment]
Closing as is addressed by https://github.com/hashicorp/consul/pull/20945, this will be released in 1.18.2 likely towards the end of next month.
